### ISSUE TITLE
Mismatch OS

### ISSUE THREAD

**Issue by gprovencher-cdpq** [dmlc/xgboost#6709]
Hi,

I observe that the rankings obtained using XGBRanker are not the same whether I use Windows or Linux. (I use xgboost 1.3.3)

`from xgboost import XGBRanker
import numpy as np
import pandas as pd

# Training set.
seed = 1
np.random.seed(seed)
input_df = pd.DataFrame({
    "feature_1": np.random.rand(1000),
    "feature_2": np.random.rand(1000)
})

# Target column has a simple relation with respect to the feature columns.
input_df["target"] = input_df["feature_1"] * input_df["feature_2"]
 
feature_columns = ["feature_1", "feature_2"]
target_columns = "target"
    
estimator = XGBRanker(
        base_score=0.5, booster="gbtree", colsample_bylevel=1.0,
        colsample_bynode=1.0, colsample_bytree=1.0, gamma=0.5, gpu_id=-1,
        importance_type='gain', interaction_constraints='',
        learning_rate=0.3, max_delta_step=1, max_depth=3,
        min_child_weight=5, missing=np.nan, monotone_constraints='()',
        n_estimators=50, n_jobs=1, num_parallel_tree=1,
        random_state=0, reg_alpha=0, reg_lambda=1.0,
        scale_pos_weight=None, subsample=1.0, tree_method="exact",
        validate_parameters=False, verbosity=0, nthread=1
)
 
X = input_df[feature_columns].copy()
y = input_df[target_columns].copy()
group = np.array([len(X)])
 
estimator.fit(X=X, y=y, group=group, sample_weight=None, base_margin=None, 
              eval_set=None, sample_weight_eval_set=None, 
              eval_group=None, eval_metric=None, 
              early_stopping_rounds=None, verbose=False, xgb_model=None, 
              feature_weights=None, callbacks=None)

print("Feature importance:", estimator.feature_importances_)

# Dataset for predictions.
np.random.seed(seed + 1)
to_predict_df = pd.DataFrame({
    "feature_1": np.random.rand(1000),
    "feature_2": np.random.rand(1000)
})
to_predict_df["result"] = estimator.predict(data=to_predict_df)

to_predict_df["answer"] = to_predict_df["feature_1"] * to_predict_df["feature_2"]
to_predict_df["answer_rank"] = to_predict_df["answer"].rank().astype(int)
to_predict_df["result_rank"] = to_predict_df["result"].rank().astype(int)
to_predict_df["rank_error"] = abs(to_predict_df["result_rank"] - to_predict_df["answer_rank"])
 
print("Mean rank error:", to_predict_df["rank_error"].sum() / len(to_predict_df))`

My understanding is that nothing in this computation should involve randomness: Am I wrong?

Thanks,
Guillaume

**Comment by trivialfis** [comment]
Could you please confirm the generated data is the same, and share the difference between output?

**Comment by gprovencher-cdpq** [comment]
I confirm the generated data is exactly the same.

**Comment by trivialfis** [comment]
There's many contributing factors to floating point difference, like different hardware, different compiler optimisations etc.   It's quite difficult to get portable result if possible at all.   I will look into it later to see if there's any human error on xgboost s side.  But I can't promise to get the exact result for 2 platforms.

**Comment by gprovencher-cdpq** [comment]
I agree that it's not possible to get the exact same results, up to machine precision. Here however something doesn't seem quite right.

When I take the same code snippet as above but use XGBRegressor instead of XGBRanker (and remove any superfluous parameters), I get the same results (8th decimal place), that is the same feature importance and prediction results, on Windows and Linux.

My concern is: which implementation gives the "correct" results?

Again, thanks for you help!
